horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/RunServiceTest.java

```diff
@@ -59,24 +75,72 @@
    EventBus eventBus;
 
    @org.junit.jupiter.api.Test
-   public void testDataSetCreated(TestInfo info) throws InterruptedException {
+   public void testStraightforwardTransformationEmptyResult(TestInfo info) throws InterruptedException {
+      BlockingQueue<DataSet> dataSetQueue = eventConsumerQueue(DataSet.class, DataSet.EVENT_NEW);
+      NamedJsonPath path = createExampleNamedJsonPath("foo", "$.value");
+      Schema schema = createExampleSchema(info, true);
+
+      assertNotNull(schema.id);
+      Transformer transformer = createExampleTransformerWithJsonPath("acme", schema, path);
+      Test exampleTest = createExampleTest(getTestName(info));
+      Test test = createTest(exampleTest);
+      addTransformer(test, transformer);
+      uploadRun("{\"corporation\":\"acme\"}", test.name);
+
+      DataSet event = dataSetQueue.poll(10, TimeUnit.SECONDS);
+      assertNotNull(event);
+      JsonNode node = event.data;
+      assertTrue(node.isArray());
+      assertTrue(node.isEmpty());
+   }
+
+   private NamedJsonPath createExampleNamedJsonPath(String name, String path) {
+      NamedJsonPath jsonpath = new NamedJsonPath();
+      jsonpath.name = name;
+      jsonpath.jsonpath = path;
+      return jsonpath;
+   }
+
+   @org.junit.jupiter.api.Test
+   public void testTransformationWithoutSchema(TestInfo info) throws InterruptedException {
       BlockingQueue<DataSet> dataSetQueue = eventConsumerQueue(DataSet.class, DataSet.EVENT_NEW);
       Test test = createTest(createExampleTest(getTestName(info)));
-      Schema schema = createExampleSchema(info);
       setTestVariables(test, "Value", "value");
 
-      uploadRun(runWithValue(schema, 42).toString(), test.name);
+      ObjectNode jsonWithoutSchema = JsonNodeFactory.instance.objectNode()
+         .put("meaning of life", "42");
+      uploadRun( "{ \"foo\":\"bar\"}", test.name);
 
       try (CloseMe ignored = roleManager.withRoles(em, Arrays.asList(TESTER_ROLES))) {
          DataSet event = dataSetQueue.poll(10, TimeUnit.SECONDS);
          assertNotNull(event);
-         assertNotNull(event.id);
-         Integer id = event.id;
-         DataSet ds = DataSet.findById(id);
-         assertNotNull(ds);
+         JsonNode node = event.data;
+         assertNotNull(node);
+         assertTrue(node.isArray());
+         assertTrue(node.isEmpty());
       }
    }
 
+   @org.junit.jupiter.api.Test
+   public void testTransformationWithoutExtractors(TestInfo info) throws InterruptedException {
+      BlockingQueue<DataSet> dataSetQueue = eventConsumerQueue(DataSet.class, DataSet.EVENT_NEW);
+      NamedJsonPath path = createExampleNamedJsonPath("foo", "$.value");
+      Schema schema = createExampleSchema(info, false);
+
+      assertNotNull(schema.id);
+      Transformer transformer = createExampleTransformerWithJsonPath("acme", schema, path);
```

Comment:
It took me a bit of reasoning to check if this test is correct - maybe it would make sense to add some comment with discussion:

This the Test has Transformer for this schema defined, and the Run contains this schema. Therefore, the input for this transformer should be `{ "foo": null }` and since it does not have a function defined the result is identity: `{ "foo": null}`. The problem here is that the `$schema` is not defined and the `Transformer.targetSchemaUri` is not set either, therefore the result of the transformation is ignored and the DataSet is really empty.

Dropping the Transformer result should be explained in the log and UI could indicate it. However I wonder if we should keep `targetSchemaUri` optional - we could rather treat it as a mandatory default value for output schema. WDYT?

Reply:
I've edited the comment above, sorry for confusion.

Reply:
The question is what should we do if we get a scalar result (like `null`, `42` or `"foobar"`) and the `targetSchemaUri` **is** set. Just failing is cowardly - maybe we could auto-promote to an object that will look like `{ "$schema": "<targetSchemaUri", "value": 42 }`

Reply:
I came across an issue yesterday that might make this question moot. Setting the schema in the function cannot be parsed. The function result does not quote the $schema name. Which fails to be converted into a JsonNode object.

> {$schema: "urn:AcmeInc:AcmeRrUs:1.0", outcome: {a: undefined}}
> 
> Failed to parse into JSON: {$schema: "urn:AcmeInc:AcmeRrUs:1.0", outcome: {a: undefined}}: com.fasterxml.jackson.core.JsonParseException: Unexpected character ('$' (code 36)): was expecting double-quote to start field name
>  at [Source: (String)"{$schema: "urn:AcmeInc:AcmeRrUs:1.0", outcome: {a: undefined}}"; line: 1, column: 3]

 Besides that I'd say setting a targetSchemaUri as a step after the function is more likely to be reliable.

Reply:
There are 2 problems with the example object in the error:
1) you need to quote `$schema` when defining the object as it contains the `$` character
2) `undefined` is not a valid JSON value. It's a valid JavaScript identifier, but not in JSON.

Reply:
The  function has the schema key quoted already. What you see here is the output of the function. Ignore the outcome  please. I was pointing out what's broken with the output.
`{$schema: "urn:AcmeInc:AcmeRrUs:1.0"   `

Reply:
I don't get an exception. I pasted the quoted text from the [result](https://github.com/whitingjr/Horreum/blob/issue-%23148-apply-transformers/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java#L1099) of processing the function.

Reply:
> The question is what should we do if we get a scalar result (like `null`, `42` or `"foobar"`) and the `targetSchemaUri` **is** set. Just failing is cowardly - maybe we could auto-promote to an object that will look like
> 
> ```
> {
>   "$schema": "<targetSchemaUri>",
>   "value": 42
> }
> ```

I have added changes to do exactly this. Making this an object node with the schema field set.